File: Sort by Date/007 -2018apple January 24 - Feb 6 2018/Sentence Reverse.java
import java.io.*;
import java.util.*;

class Solution {

  static char[] reverseWords(char[] arr) {
    
    if(arr == null || arr.length <= 1){
        return arr ; 
    }
    
    int len = arr.length ;         
     
    reverse(arr, 0, len-1);
    
    int left = 0, right = 0; 
    
    while(left < len && right < len){
      char current = arr[right];
      
      if(current == ' ' || right == len-1) //handle single word
      { //found a word
        int end = current == ' '? right - 1 : right; 
        reverse(arr, left, end);  //reverse chars in the word
        left = right+1; //move to next word
      }
      
      right++ ;           
    }
    
    return arr ; 
  }
  
  static void reverse(char[] arr, int start, int end){
    int i=start, j=end; 
    
    while(i<j){
      char temp = arr[i]; 
      arr[i] = arr[j]; 
      arr[j] = temp ;
      
      i++;
      j--;
    }
  }

  public static void main(String[] args) {
  
  }

}

File: Sort by Date/009 - cat Feb 25 - March 29 2018/Sentence reverse.cs
using System;

class Solution
{
    public static char[] ReverseWords(char[] arr)
    {
        if (arr == null || arr.Length == 0)
            return new char[0];

        var length = arr.Length;

        reverse(arr, 0, length - 1);

        int start = 0;
        for (int i = 0; i < length; i++)
        {
            var current = arr[i];

            var isSpace = current == ' ';
            if (isSpace)
                continue;

            if (i == 0 || arr[i - 1] == ' ')
                start = i;

            if (i == length - 1 || arr[i + 1] == ' ')
            {
                // reverse the word
                reverse(arr, start, i);
            }
        }

        return arr;
    }

    private static void reverse(char[] arr, int start, int end)
    {
        int left = start;
        int right = end;

        while (left < right)   // diff = left - start -> determine: end - left
        {
            var tmp = arr[left];
            arr[left] = arr[right];
            arr[end - (left - start)] = tmp;

            left++;
        }
    }

    static void Main(string[] args)
    {
        var reversed = ReverseWords(new char[] { 'a', ' ', 'b', ' ', 'c', 'a', 't' });

        foreach (var item in reversed)
        {
            Console.WriteLine(item);
        }
    }
}

/*
reverse(char[] arr, int start, int end)
  
"a b cat"  -> "tac b a" -> "cat b a"  -> Pattern 
Iterate the reversed array -> if word is found

cat
  |
  next index = ' ' or it is last index in the array
  */